File: python/Converters/test/tConvert.cc
//# tConvert.cc: Test program for libpython's C++/Python converters
//# Copyright (C) 2006
//# Associated Universities, Inc. Washington DC, USA.
//#
//# This library is free software; you can redistribute it and/or modify it
//# under the terms of the GNU Library General Public License as published by
//# the Free Software Foundation; either version 2 of the License, or (at your
//# option) any later version.
//#
//# This library is distributed in the hope that it will be useful, but WITHOUT
//# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
//# FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Library General Public
//# License for more details.
//#
//# You should have received a copy of the GNU Library General Public License
//# along with this library; if not, write to the Free Software Foundation,
//# Inc., 675 Massachusetts Ave, Cambridge, MA 02139, USA.
//#
//# Correspondence concerning AIPS++ should be addressed as follows:
//#        Internet email: aips2-request@nrao.edu.
//#        Postal address: AIPS++ Project Office
//#                        National Radio Astronomy Observatory
//#                        520 Edgemont Road
//#                        Charlottesville, VA 22903-2475 USA
//#
//# $Id: tConvert.cc,v 1.4 2006/11/06 00:14:44 gvandiep Exp $

#include <casacore/python/Converters/PycExcp.h>
#include <casacore/python/Converters/PycBasicData.h>
#include <casacore/python/Converters/PycValueHolder.h>
#include <casacore/python/Converters/PycRecord.h>
#include <casacore/python/Converters/PycArray.h>
#include <casacore/casa/Arrays/ArrayIO.h>
#include <casacore/casa/BasicSL/STLIO.h>

#include <boost/python.hpp>

using namespace boost::python;

namespace casacore { namespace python {

  struct TConvert
  {
    TConvert() {}
    Bool  testbool (Bool in)
      {cout << "bool " << in << endl; return in;}
    Int testint (Int in)
      {cout << "Int " << in << endl; return in;}
    Int64 testint64 (Int64 in)
      {cout << "Int64 " << in << endl; return in;}
    Int testssize (::ssize_t in)
      {cout << "ssize " << in << endl; return in;}
    Float testfloat (Float in)
      {cout << "Float " << in << endl; return in;}
    Double testdouble (Double in)
      {cout << "Double " << in << endl; return in;}
    Complex testcomplex (const Complex& in)
      {cout << "Complex " << in << endl; return in;}
    DComplex testdcomplex (const DComplex& in)
      {cout << "DComplex " << in << endl; return in;}
    String teststring (const String& in)
      {cout << "String " << in << endl; String out=in; return out;}
    Record testrecord (const Record& in)
      {cout << "Record "; in.print(cout); cout << endl; return in;}
    ValueHolder testvh (const ValueHolder& in)
      {cout << "VH " << in.dataType() << endl; return in;}
    Vector<Bool> testvecbool (const Vector<Bool>& in)
      {cout << "VecBool " << in << endl; return in;}
    Vector<Int> testvecint (const Vector<int>& in)
      {cout << "VecInt " << in << endl; return in;}
    Vector<DComplex> testveccomplex (const Vector<DComplex>& in)
      {cout << "VecComplex " << in << endl; return in;}
    Vector<String> testvecstr (const Vector<String>& in)
      {cout << "VecStr " << in << endl; return in;}
    std::vector<bool> teststdvecbool (const std::vector<bool>& in)
      {cout << "vecbool " << in << endl; return in;}
    std::vector<uInt> teststdvecuint (const std::vector<uInt>& in)
      {cout << "vecuInt " << in << endl; return in;}
    std::vector<std::vector<uInt> > teststdvecvecuint
    (const std::vector<std::vector<uInt> >& in)
      {cout << "vecvecuInt " << in << endl; return in;}
    std::vector<ValueHolder> teststdvecvh (const std::vector<ValueHolder>& in)
      {cout << "vecvh " << in.size() << endl; return in;}
    IPosition testipos (const IPosition& in)
      {cout << "IPos " << in << endl; return in;}
  };


  void testConvert()
  {
    class_<TConvert> ("tConvert", init<>())
      .def ("testbool",       &TConvert::testbool)
      .def ("testint",        &TConvert::testint)
      .def ("testint64",      &TConvert::testint64)
      .def ("testssize",      &TConvert::testssize)
      .def ("testfloat",      &TConvert::testfloat)
      .def ("testdouble",     &TConvert::testdouble)
      .def ("testcomplex",    &TConvert::testcomplex)
      .def ("testdcomplex",   &TConvert::testdcomplex)
      .def ("teststring",     &TConvert::teststring)
      .def ("testrecord",     &TConvert::testrecord)
      .def ("testvh",         &TConvert::testvh)
      .def ("testvecbool",    &TConvert::testvecbool)
      .def ("testvecint",     &TConvert::testvecint)
      .def ("testveccomplex", &TConvert::testveccomplex)
      .def ("testvecstr",     &TConvert::testvecstr)
      .def ("teststdvecbool", &TConvert::teststdvecbool)
      .def ("teststdvecuint", &TConvert::teststdvecuint)
      .def ("teststdvecvecuint", &TConvert::teststdvecvecuint)
      .def ("teststdvecvh"  , &TConvert::teststdvecvh)
      .def ("testipos",       &TConvert::testipos)
      ;
  }

}}


BOOST_PYTHON_MODULE(_tConvert)
{
  // Register the required converters.
  casacore::python::register_convert_excp();
  casacore::python::register_convert_basicdata();
  casacore::python::register_convert_casa_valueholder();
  casacore::python::register_convert_casa_record();
  casacore::python::register_convert_std_vector<bool>();
  casacore::python::register_convert_std_vector<casacore::uInt>();
  casacore::python::register_convert_std_vector<std::vector<casacore::uInt> >();
  casacore::python::register_convert_std_vector<casacore::ValueHolder>();

  // Execute the test.
  casacore::python::testConvert();
}

File: python/Converters/test/tConvert.py
#!/usr/bin/env python
from _tConvert import *

def dotest(t):

    print ''
    print 'begin dotest'
    print t.testbool (True);
    print t.testbool (False);
    print t.testint (-1);
    print t.testint (10L);
    print t.testint64 (-123456789013L);
    print t.testint64 (123456789014L);
    print t.testssize (-2);
    print t.testssize (11);
    print t.testfloat (3.14);
    print t.testfloat (12);
    print t.teststring ("this is a string");

    print t.testipos ([2,3,4]);
    print t.testipos (1);
    print t.testipos (NUM.array([2]));
    print t.testipos (NUM.array(3));

    print t.testvecbool ([True,False,False,True])
    print t.testvecint ([1,2,3,4]);
    print t.testvecint ([]);
    print t.testvecint ((-1,-2,-3,-4));
    print t.testvecint (-10);
    print t.testvecint (NUM.array((10,11,12)));
    print t.testvecint (NUM.array(1));
    print t.testveccomplex ([1+2j, -1-3j, -1.5+2.5j]);
    print t.testvecstr (["a1","a2","b1","b2"])
    print t.testvecstr (())
    print t.testvecstr ("sc1")
    print t.teststdvecbool ([False,True])
    print t.teststdvecuint ([1,2,4])
    print t.teststdvecuint (())
    print t.teststdvecuint (10)
    print t.teststdvecvecuint ([[1,2,4]])
    print t.teststdvecvecuint ((()))
    print t.teststdvecvecuint (())
    print t.teststdvecvecuint ([1,2,4])
    print t.teststdvecvecuint (20)

    print t.testvh (True);
    print t.testvh (2);
    print t.testvh (1234567890123L);
    print t.testvh (1.3);
    print t.testvh (10-11j);
    print t.testvh ("str");
    print t.testvh ([True]) + 0;         # add 0 to convert numpy to integer
    print t.testvh ([2,4,6,8,10]);
    print t.testvh ([1.3,4,5,6]);
    print t.testvh ([10-11j,1+2j]);
#    print t.testvh ([]);
    print t.testvh (["str1","str2"]);
    print t.testvh ({"shape":[2,2],"array":["str1","str2","str3","str4"]});
    a  =  t.testvh ({"shape":[2,2],"array":["str1","str2","str3","str4"]});
    print a;
    print t.testvh (a);

    a  =  t.testvh ([10-11j,1+2j]);
    print a.shape;
    print t.testvh (a);

    b  =  NUM.int32([[2,3],[4,5]]);
    print b;
    print t.testvh (b);

    b  =  NUM.int32([1,2,3,4,5,6,7,8,9,10]);
    print b[2:9:2];
    print t.testvh (b[2:9:2]);

    b  =  NUM.array([1,2,3,4,5,6,7,8,9,10.]);
    print b[2:9:2];
    print t.testvh (b[2:9:2]);
    a = b[2:9:2];
    print t.testvh (a);

    print t.testvh(NUM.array([20.+10j]));
    print t.testvh(NUM.array([21.]));
    print t.testvh(NUM.array(21.));

    print '>>>';
    res = t.testvh (NUM.array([]));
    print '<<<';
    print res.shape;
    print '>>>';
    res = t.testvh (NUM.array([[]]));
    print '<<<';
    print res.shape;

    # Test a sequence of ValueHolders
    print t.teststdvecvh([2, 1.3, [True,False]]);

    # On 64-bit machines the output also contains 'dtype=int32'
    # So leave it out.
    a = t.testrecord({"int":1, "int64":123456789012L, "str":"bc", 'vecint':[1,2,3]})
    print '>>>'
    print a
    print '<<<'
    print 'end dotest'
    print ''


def testarrvh(arr):
    print '    testarrvh';
    print t.testvh(arr);
    print t.testvh(arr[0]);
    print t.testvh([arr[0]]);
    print t.testvh([arr[0], arr[1]]);

def testarrb(arr):
    testarrvh(arr);
    print t.testbool(arr[0]);

def testarri(arr):
    testarrvh(arr);
    print t.testint(arr[0]);
    print t.testint64(arr[0]);
    print t.testssize(arr[0]);
    print t.testfloat(arr[0]);
    print t.testcomplex(arr[0]);

def testarrf(arr):
    testarrvh(arr);
    print t.testfloat(arr[0]);
    print t.testcomplex(arr[0]);

def testarrc(arr):
    testarrvh(arr);
    print t.testcomplex(arr[0]);

def testnps():
    testarrb(NUM.array([True,False]));
    testarri(NUM.int8([-6,-7]));
    testarri(NUM.uint8([5,6]));
    testarri(NUM.int16([-16,-17]));
    testarri(NUM.uint16([15,16]));
    testarri(NUM.int32([-26,-27]));
    testarri(NUM.uint32([25,26]));
    testarri(NUM.int64([-36,-37]));
    testarri(NUM.uint64([35,36]));
    testarrf(NUM.float32([-46,-47]));
    testarrf(NUM.float64([45,46]));
    testarrc(NUM.complex64([-56-66j,-57-67j]));
    testarrc(NUM.complex128([-76-86j,-77-87j]));

def testnp():
    # Test byte and sbyte.
    b = NUM.int8([-1,-2]);
    print t.testvh(b);
    b = NUM.uint8([211,212]);
    print t.testvh(b);
    print '>>>';
    res = t.testvh(NUM.array([]));
    print '<<<';
    print res.shape;
    print t.testvh(NUM.array([["abcd","c"],["12","x12"]]));
    testnps();

if __name__ == "__main__":

    import numpy as NUM;
    t = tConvert();
    print "Doing numpy/array test ..."
    testnp();
    # Do other tests.
    dotest(t)
